lib/puppet/ssl/ssl_provider.rb

```diff
@@ -0,0 +1,192 @@
+require 'puppet/ssl'
+
+# SSL Provider creates `SSLContext` objects that can be used to create
+# secure connections.
+#
+# @api private
+class Puppet::SSL::SSLProvider
+  # Create an insecure `SSLContext`. Connections made from the returned context
+  # will not authenticate the server, i.e. `VERIFY_NONE`, and are vulnerable to
+  # MITM. Do not call this method.
+  #
+  # @return [Puppet::SSL::SSLContext] A context to use to create connections
+  # @api private
+  def create_insecure_context
+    store = create_x509_store([], [], false)
+
+    Puppet::SSL::SSLContext.new(store: store, verify_peer: false)
+  end
+
+  # Create an `SSLContext` using the trusted `cacerts` and optional `crls`.
+  # Connections made from the returned context will authenticate the server,
+  # i.e. `VERIFY_PEER`, but will not use a client certificate.
+  #
+  # The `crls` parameter must contain CRLs corresponding to each CA in `cacerts`
+  # depending on the `revocation` mode. See {#create_context}.
+  #
+  # @param cacerts [Array<OpenSSL::X509::Certificate>] Array of trusted CA certs
+  # @param crls [Array<OpenSSL::X509::CRL>] Array of CRLs
+  # @param revocation [:chain, :leaf, false] revocation mode
+  # @return [Puppet::SSL::SSLContext] A context to use to create connections
+  # @raise (see #create_context)
+  # @api private
+  def create_root_context(cacerts:, crls: [], revocation: Puppet[:certificate_revocation])
```

Comment:
More of curiosity - would this be used in current agent <-> server interactions where we always require client authentication? Or is it meant for completeness in case there is a need to communicate with some endpoint with only server authentication?

Reply:
It's for the case where we don't have a client cert yet, during the agent bootstrap process.

There will be a separate PR post-6.4.0 so that puppet can create an ssl_context that trusts the system store, as is needed to download file sources from github, send reports to non-puppetservers, etc.